### PR TITLE
Changed #tfop signature to always return TensorHandle

### PR DESCRIPTION
**Pull request by mhong**
(some TF ops also return resource handles and variant handles, which can be supported later).

The APIs in RawOps (and other APIs in TensorFlow stdlib) will continue to consume
and produce Tensor's, as Tensor is the user-facing class.

A #tfop can return one or a tuple of TensorHandle/[TensorHandle] elements, as
illustrated in the following example in the generated code. In this case, the
#tfop is returning a tuple of 3 elements, where the first element is a tensor
list.

```
@inlinable @inline(__always)
public static func batch<T: AccelerableByTensorFlow>(
  inTensors: [Tensor<T>],
  numBatchThreads: Int64,
  maxBatchSize: Int64,
  maxEnqueuedBatches: Int64 = 10,
  batchTimeoutMicros: Int64,
  allowedBatchSizes: [Int32],
  gradTimeoutMicros: Int64,
  container: String,
  sharedName: String,
  batchingQueue: String
) -> (batchedTensors: [Tensor<T>], batchIndex: Tensor<Int64>, id: Tensor<Int64>) {
  let ret: (batchedTensors: [TensorHandle<T>], batchIndex: TensorHandle<Int64>, id: TensorHandle<Int64>) = #tfop("Batch",
    inTensors,
    num_batch_threads: numBatchThreads,
    max_batch_size: maxBatchSize,
    max_enqueued_batches: maxEnqueuedBatches,
    batch_timeout_micros: batchTimeoutMicros,
    allowed_batch_sizes: allowedBatchSizes,
    grad_timeout_micros: gradTimeoutMicros,
    container: container,
    shared_name: sharedName,
    batching_queue: batchingQueue)
  return (ret.0.map(Tensor.init), Tensor(handle: ret.1), Tensor(handle: ret.2))
}
```

### CODE REVIEW

**Comment by mhong** [comment]
@LaurentMazare I haven't written much python code for serious projects, so I'd welcome any feedback on style, etc. Thanks!

File: generate_wrappers.py [hunk 1]
```diff
@@ -416,17 +428,57 @@ def generate_code(op, api_def, enum_store):
...
+    body = 'return #tfop({tfop_args})'.format(tfop_args=tfop_args)
+  elif len(return_name_and_types) >= 1:
+    # Example body with 1 return tensor:
+    # let ret: [TensorHandle<Int32>] = #tfop("ConcatOffset",
+    #   concatDim,
+    #   shape)
+    # return ret.0.map(Tensor.init)
+    #
+    # Example body with 2 return tensors:
+    # let ret: (loss: Tensor<T>, backprop: Tensor<T>) = #tfop("SoftmaxCrossEntropyWithLogits",
```
Comment (rxwei) [inline generate_wrappers.py]: The type signature is incorrect here, and the code example in this comment won't work. 

1. Tuple elements should be handles, not tensors.

2. If you add specific labels to the tuple type signature (like in your example `: (loss: Tensor<T>, backprop: Tensor<T>)`), then you'll need to access elements of `ret` via `ret.loss` and `ret.backprop`. Your examples below (`ret.0`, etc) won't work. So don't add labels to the type signature - this makes the compiler erase the label information so that you won't need labels to access them (you can then just use `ret.0`, `ret.1`, etc).

Comment (mhong) [reply generate_wrappers.py]: Fixed this stale comment. Thanks for the catch!

The real code from RawOpsGenerated.swift (part of this PR) is:
```swift
@inlinable @inline(__always)
public static func softmaxCrossEntropyWithLogits<T: BinaryFloatingPoint>(
  features: Tensor<T>,
  labels: Tensor<T>
) -> (loss: Tensor<T>, backprop: Tensor<T>) {
  let ret: (loss: TensorHandle<T>, backprop: TensorHandle<T>) = #tfop("SoftmaxCrossEntropyWithLogits",
    features,
    labels,
    T: T.self)
  return (Tensor(handle: ret.0), Tensor(handle: ret.1))
}
```

using `ret.0` works (I've tested this PR in the context of https://github.com/apple/swift/pull/17648#pullrequestreview-133425846), and I'd prefer that since it keeps the (generated) code simpler and more readable.

Comment (lattner) [reply generate_wrappers.py]: This approach is really nice.  I think it makes perfect sense for the wrapper generator to produce TensorHandle directly from #tfop's.  With this approach, can we completely remove the goop in CSApply that is handling Tensor results?

Comment (mhong) [reply generate_wrappers.py]: @lattner Does https://github.com/apple/swift/pull/17648 do what you have in mind?

Comment (rxwei) [reply generate_wrappers.py]: Ok, turns out index access also works for named tuple elements. One question though:

```swift
let ret: (loss: TensorHandle<T>, backprop: TensorHandle<T>)
```

Why are element labels in the tuple type signature necessary? They are not type-checked by the `#tfop` type checker, and the are making they generated code *less* simple and *less* readable because they are unnecessarily long and you aren't using element labels to access elements anyway.

If you remove the tuple element labels in the type signature of `ret`
```swift
let ret: (TensorHandle<T>, TensorHandle<T>) = #tfop("SoftmaxCrossEntropyWithLogits",
  features,
  labels,
  T: T.self)
return (Tensor(handle: ret.0), Tensor(handle: ret.1))
```

This is arguably shorter and much simpler.

Comment (dan-zheng) [reply generate_wrappers.py]: I agree with removing tuple element labels from `#tfop` for simplicity. Labels should be preserved on the actual wrapper `func`.

File: generate_wrappers.py [hunk 2]
```diff
@@ -385,12 +396,13 @@ def generate_code(op, api_def, enum_store):
   return_name_and_types = [
       (swiftified_name(a.name), arg_def_type_as_string(a))
       for a in op.output_arg]
+  has_arrow = ' -> ' if len(return_name_and_types) > 0 else ''
```
Comment (LaurentMazare) [inline generate_wrappers.py]: Could we avoid this `has_arrow` variable by including the arrow string in `return_type` directly ?

File: generate_wrappers.py [hunk 3]
```diff
@@ -361,6 +361,17 @@ def maybe_named(name):
...
+# Two cases based on if handle_var is TensorHandle or [TensorHandle]:
+# 1. Case TensorHandle: h -> Tensor(handle: h)
+# 2. Case [TensorHandle]: h -> h.map(Tensor.init)
+def convert_handle_to_tensor(handle_var, is_list):
+  if is_list:
+    return handle_var + '.map(Tensor.init)'
+  return 'Tensor(handle: ' + handle_var + ')'
+
+# If type_name is [TensorHandle<T>], it is a list type.
+def type_is_list(type_name):
```
Comment (LaurentMazare) [inline generate_wrappers.py]: The implementation seems a bit fragile, maybe we could make this more robust by including in return_name_and_types whether this return value has a list type.

File: generate_wrappers.py [hunk 4]
```diff
@@ -416,17 +428,57 @@ def generate_code(op, api_def, enum_store):
           input_names_and_types +
           attr_names_and_types +
           missing_types)]
+  body = ''
+  if len(return_name_and_types) == 0:
+    body = 'return #tfop({tfop_args})'.format(tfop_args=tfop_args)
```
Comment (LaurentMazare) [inline generate_wrappers.py:433]: Do we really need the `return` bit in the body in this case as there are no actual output ?

**Review by LaurentMazare — COMMENTED** [review]
Thanks for the PR, looks pretty neat. I've added a couple mostly cosmetic comments.

File: generate_wrappers.py (hunk 4 above)
Comment (mhong) [reply generate_wrappers.py:433]: I believe so. One example is:
```swift
@inlinable @inline(__always)
public static func abort(
  errorMsg: String,
  exitWithoutError: Bool = false
) {
  return #tfop("Abort",
    error_msg: errorMsg,
    exit_without_error: exitWithoutError)
}
```

If we don't do return, type checker cannot infer the return type of `#tfop` in this context. 

Also, note the above generated swift code did not change in this PR -- the python code here is a refactoring.

File: generate_wrappers.py (hunk 3 above)
Comment (mhong) [reply generate_wrappers.py]: Done.

File: generate_wrappers.py (hunk 2 above)
Comment (mhong) [reply generate_wrappers.py]: Done.

File: generate_wrappers.py (hunk 1 above)
Comment (mhong) [reply generate_wrappers.py]: Done.

**Review by LaurentMazare — APPROVED** [review]

File: generate_wrappers.py (hunk 4 above)
Comment (LaurentMazare) [reply generate_wrappers.py:433]: Got it, sorry for the misleading comment.

**Review by rxwei — APPROVED** [review]

**Comment by mhong** [comment]
Thank you all for the review!
